shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -62,15 +62,57 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
       env: MutMap[Str, TypeInfo],
       mthEnv: MutMap[(Str, Str) \/ (Opt[Str], Str), MethodType],
       lvl: Int,
+      quoteSkolemEnv: MutMap[Str, SkolemTag], // * SkolemTag for variables in quasiquotes
+      freeVarsInCurrentQuote: MutSet[ST], // * Free variables appearing in the current quote scope
+      inQuote: Bool, // * Is in quasiquote
       inPattern: Bool,
       tyDefs: Map[Str, TypeDef],
       tyDefs2: MutMap[Str, DelayedTypeInfo],
       inRecursiveDef: Opt[Var], // TODO rm
       extrCtx: ExtrCtx,
   ) {
-    def +=(b: Str -> TypeInfo): Unit = env += b
+    def +=(b: Str -> TypeInfo): Unit = {
+      env += b
+      /**
+        * For quoted lambdas and let bindings, `quoteSkolemEnv` should not have bindings with the same name in one quoted context
+        * because we call `enterQuotedScope` function that resets the `quoteSkolemEnv`.
+        * This check (i.e., `!quoteSkolemEnv.contains(b._1)`) is for quoted UCS desugared terms that contain bindings
+        * and it is safe to share the same skolem tag since the temporary variables are always bound.
+        */
+      if (inQuote && !quoteSkolemEnv.contains(b._1)) {
+        val tag = SkolemTag(freshVar(NoProv, N, nameHint = S(b._1))(lvl))(NoProv)
+        println(s"Create skolem tag $tag for ${b._2} in quasiquote.")
+        quoteSkolemEnv += b._1 -> tag
+      }
+    }
     def ++=(bs: IterableOnce[Str -> TypeInfo]): Unit = bs.iterator.foreach(+=)
-    def get(name: Str): Opt[TypeInfo] = env.get(name) orElse parent.dlof(_.get(name))(N)
+    def get(name: Str): Opt[TypeInfo] = {
+      /**
+        * 1. If we try to get a quoted variable from a quoted context, or get a non-quoted variable from a non-quoted context,
+        * we can directly search the `env`;
+        * 2. If we try to get a quoted variable from a non-quoted context, we need to wrap it with its contextual Skolem;
+        * 3. Otherwise, it is a top-level defined symbol, or the context has a quoted parent.
+        */
+      def rec(ctx: Ctx): Opt[TypeInfo] = {
+        if (inQuote === ctx.inQuote) ctx.env.get(name) orElse ctx.parent.flatMap(rec(_))
+        else if (!inQuote) (ctx.env.get(name), ctx.getQuoteSkolem(name)) match {
+          case (S(VarSymbol(ty, _)), S(tag)) =>
+            S(VarSymbol(TypeRef(TypeName("Var"), ty :: tag :: Nil)(noProv), Var(name)))
+          case _ => ctx.parent.flatMap(rec(_))
+        }
+        else ctx.parent match {
+          case S(parent) => rec(parent)
+          case _ => ctx.env.get(name)
+        }
+      }
+      rec(this)
+    }
+    def getQuoteSkolem(name: Str): Opt[SkolemTag] = quoteSkolemEnv.get(name) orElse parent.dlof(_.getQuoteSkolem(name))(N)
+    def getCtxTy: ST = freeVarsInCurrentQuote.foldLeft[ST](BotType)((res, ty) => res | ty)
+    def traceFV(fv: ST): Unit = {
```

Comment:
We've already concluded in a previous meeting that you meant to use "track" here, not "trace".

Reply:
Also you're sometimes calling this to track SEVERAL variables.
```suggestion
    def trackFVs(fvsType: ST): Unit = {
```